clang/lib/Sema/SemaTemplateDeduction.cpp

```diff
@@ -859,8 +861,10 @@ class PackDeductionScope {
     // by this pack expansion, then clear out the deduction.
     DeducedFromEarlierParameter = !Deduced[Index].isNull();
     DeducedPack Pack(Index);
-    Pack.Saved = Deduced[Index];
-    Deduced[Index] = TemplateArgument();
+    if (!FinishingDeduction) {
+      Pack.Saved = Deduced[Index];
+      Deduced[Index] = TemplateArgument();
+    }
```

Comment:
A comment describing what `FinishingDeducing` is for would be useful. It looks like the intent is that in this mode we just gather information and don't update `Deduced` at all; if so, there's an update to `Deduced` in `finishConstruction` that you might have missed.

Reply:
Will do.

Yeah that's exactly it. The idea is to reuse the parts of the PackScope which help iterating a pack, but in a context where we are not producing a deduction, just checking a previous deduction.

Reply:
Can we add pack-related tests?

Reply:
Note we aren't doing any pack specific changes here, this is just for better code reuse.

We have tons of tests in the test suite with partial ordering of function templates which involve packs.

One example I can give offhand (because I mentioned to @zygoloid on another thread) is 'temp_deduct_type_example3' in `CXX/CXX/drs/cwg6xx.cpp`:
```C++
   template<class T, class... U> void f(T*, U...){}
    template<class T> void f(T){}
    template void f(int*);
```
Do you have something in mind which might be missing?